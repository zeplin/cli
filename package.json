{
  "name": "@zeplin/cli",
  "version": "1.1.13",
  "description": "Zeplin CLI",
  "main": "./dist/src/index.js",
  "types": "./dist/types/index",
  "bin": {
    "zeplin": "bin/cli.js"
  },
  "files": [
    "dist",
    "bin"
  ],
  "scripts": {
    "test": "jest",
    "build:dev": "tsc --sourceMap",
    "build": "tsc && npm run prepare-api",
    "build:clean": "rm -rf dist/ && npm run build",
    "lint": "eslint --ext .js,.ts .",
    "prepare": "npm run build:clean",
    "prepare-api": "api-extractor run",
    "generate-docs": "api-documenter markdown --input-folder temp/ --output-folder docs/"
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run lint"
    }
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/zeplin/cli.git"
  },
  "bugs": {
    "url": "https://github.com/zeplin/cli/issues"
  },
  "homepage": "https://github.com/zeplin/cli#readme",
  "devDependencies": {
    "@microsoft/api-documenter": "^7.8.0",
    "@microsoft/api-extractor": "^7.8.0",
    "@types/ci-info": "^2.0.0",
    "@types/express": "^4.17.2",
    "@types/fs-extra": "^8.0.0",
    "@types/git-url-parse": "^9.0.0",
    "@types/hapi__joi": "^17.1.4",
    "@types/inquirer": "^6.5.0",
    "@types/jest": "^25.2.1",
    "@types/jsonwebtoken": "^8.3.4",
    "@types/node": "^12.7.5",
    "@types/parse-git-config": "^3.0.0",
    "@types/semver": "^7.3.4",
    "@types/strip-comments": "^2.0.0",
    "@types/update-notifier": "^4.1.0",
    "@types/url-join": "^4.0.0",
    "@types/winston": "^2.4.4",
    "@typescript-eslint/eslint-plugin": "^2.3.0",
    "@typescript-eslint/parser": "^2.3.0",
    "@zeplin/eslint-config": "^2.2.0",
    "eslint": "^6.4.0",
    "eslint-import-resolver-typescript": "^1.1.1",
    "husky": "^4.2.5",
    "jest": "^25.5.4",
    "ts-jest": "^25.5.1",
    "typescript": "^4.4.4"
  },
  "dependencies": {
    "@hapi/joi": "^17.1.1",
    "axios": "^0.21.4",
    "chalk": "^3.0.0",
    "chokidar": "^3.3.1",
    "ci-info": "^2.0.0",
    "commander": "^6.2.0",
    "cosmiconfig": "^7.0.0",
    "execa": "^5.0.0",
    "express": "^4.17.1",
    "fs-extra": "^8.1.0",
    "git-url-parse": "^11.4.3",
    "has-yarn": "^2.1.0",
    "http-status-codes": "^1.4.0",
    "import-from": "^3.0.0",
    "import-local": "^3.0.2",
    "inquirer": "^7.0.0",
    "inquirer-fuzzy-path": "^2.3.0",
    "inquirer-search-checkbox": "^1.0.0",
    "inquirer-search-list": "^1.2.6",
    "is-installed-globally": "^0.3.2",
    "is-yarn-global": "^0.3.0",
    "jsonwebtoken": "^8.5.1",
    "latest-version": "^5.1.0",
    "mask-json": "^1.0.3",
    "open": "^7.3.0",
    "ora": "^5.1.0",
    "parse-git-config": "^3.0.0",
    "semver": "^5.7.1",
    "strip-ansi": "^6.0.0",
    "strip-comments": "^2.0.1",
    "ts-dedent": "^1.1.0",
    "update-notifier": "^4.0.0",
    "url-join": "^4.0.1",
    "winston": "^3.2.1"
  }
}
