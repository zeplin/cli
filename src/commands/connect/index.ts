export * from "./connect";
export * from "./delete";
export * from "./initialize";
