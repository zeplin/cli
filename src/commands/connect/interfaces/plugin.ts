import { PrismLang } from "./prism";

/**
 * @public
 */
export interface LeveledLogMethod {
    (message: string): void;
    (message: string, meta: unknown): void;
    (message: string, ...meta: any[]): void;
    (infoObject: object): void;
}

/**
 * Contains custom plugin configuration and logger
 *
 * @public
 */
export interface PluginContext {
    /** {@inheritdoc PluginConfig} */
    config?: PluginConfig;
    /** {@inheritdoc ComponentConfig} */
    components: ComponentConfig[];
    logger: Logger;
}

/**
 * Custom created link for a component
 *
 * @public
 */
export interface Link {
    /** Name of the link */
    name?: string;
    /** Full URL to the resource */
    url: string;
    /**
     * Type of the link.
     * Required to show a pretty link for certain types of links on Zeplin.
     */
    type: LinkType;
}

/**
 * Link types for application link logo
 * Use `LinkType.custom` if you are not sure what to do
 * @public
 */
export enum LinkType {
    storybook = "storybook",
    github = "github",
    gitlab = "gitlab",
    bitbucket = "bitbucket",
    custom = "custom"
}

/**
 * Contains processed component data
 *
 * @public
 */
export interface ComponentData {
    /** Description of the component */
    description?: string;
    /** Language of the snippet for highlighting */
    lang?: PrismLang;
    /** Code snippet of the component */
    snippet?: string;
    /** {@inheritdoc Link} */
    links?: Link[];
}

/**
 * Arbitrary key/values for custom plugin configuration.
 *
 * @public
 */
export interface PluginConfig {
    [key: string]: unknown;
}

/**
 * Contains basic configuration for a component.
 *
 * @public
 */
export interface ComponentConfigBase {
    /** Path to the file, relative to project folder. */
    path: string;
    /** Zeplin component names related to this component file */
    zeplinNames?: string[];
    /** Zeplin component source IDs related to this component file */
    zeplinIds?: string[];
    /** Name for the component */
    name?: string;
}

/**
 * Use it as a value of {@link ComponentConfigCustom} keys for custom link composition
 *
 * @public
 */
export interface CustomUrlConfig {
    urlPath: string;
}

/**
 * Custom key/value configuration for a component.
 * Can be used to compose custom links for components
 * or can be processed by plugins for custom usage.
 *
 * @public
 */
export interface ComponentConfigCustom {
    [key: string]: CustomUrlConfig | unknown;
}

/**
 * Union of {@link ComponentConfigBase} and {@link ComponentConfigCustom}
 *
 * @public
 */
export type ComponentConfig = ComponentConfigBase & ComponentConfigCustom;

/**
 * Interface for Zeplin CLI Connected Components plugins
 *
 * @public
 */
export interface ConnectPlugin {
    /**
     * CLI invokes this method once the package is loaded.
     * PluginContext contains arbitrary configuration set for the plugin
     * on components config file.
     *
     * This method is optional. Implement it to initialize plugin locals etc.
     * based on plugin configuration.
     *
     *
     * @param pluginContext - {@link PluginContext}
     */
    init?(pluginContext: PluginContext): Promise<void>;

    /**
     * CLI invokes this method for each component in the configration file.
     *
     * @param componentConfig - {@link ComponentConfig}
     * @returns - {@link ComponentData}
    */
    process(componentConfig: ComponentConfig): Promise<ComponentData>;

    /**
     * CLI invokes this method for each component in the configuration file
     * to determine if this plugin should process this component
     *
     * @param componentConfig - {@link ComponentConfig}
     * @returns true if the plugin supports the component, false otherwise
    */
    supports(componentConfig: ComponentConfig): boolean;
}

/**
 * Interface for Zeplin CLI's logger.
 *
 * @public
 */
export interface Logger {
    error: LeveledLogMethod;
    warn: LeveledLogMethod;
    info: LeveledLogMethod;
    debug: LeveledLogMethod;
}
