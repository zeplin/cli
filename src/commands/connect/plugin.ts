import { ComponentConfig, ComponentData, StorybookConfig } from "connect-plugin";
import { CLIError } from "../../errors";
import {
    ConnectedComponent, Data, ComponentConfigFile, ConnectedBarrelComponents, Link, ConnectPluginModule
} from "./interfaces";
import urljoin from "url-join";
import { defaults } from "../../config/defaults";

interface ConnectPluginConstructor {
    new(): ConnectPluginModule;
}

// Helper method to initialize plugin classses
const constructConnectPlugin = (Constructor: ConnectPluginConstructor): ConnectPluginModule => new Constructor();

const removeEmptyFields = (componentData: ComponentData): ComponentData => {
    if (typeof componentData.description === "undefined" || componentData.description.trim() === "") {
        delete componentData.description;
    }

    if (typeof componentData.snippet === "undefined" || componentData.snippet.trim() === "") {
        delete componentData.snippet;
    }

    return componentData;
};

const prepareStorybookLinks = (baseUrl: string, storybookConfig: StorybookConfig): string[] => {
    const {
        kind,
        stories
    } = storybookConfig;

    const urlEncodedKind = encodeURIComponent(kind);
    if (stories) {
        return stories.map(story =>
            urljoin(baseUrl, `?selectedKind=${urlEncodedKind}&selectedStory=${encodeURIComponent(story)}`)
        );
    }

    return [urljoin(baseUrl, `?selectedKind=${kind}`)];
};

const importPlugins = async (plugins: string[]): Promise<ConnectPluginModule[]> => {
    try {
        const imports = plugins.map(async moduleName => {
            const connectPluginConstructor = (await import(moduleName)).default as ConnectPluginConstructor;

            const connectPluginInstance = constructConnectPlugin(connectPluginConstructor);
            connectPluginInstance.name = moduleName;
            return connectPluginInstance;
        });

        const pluginInstances = await Promise.all(imports);
        return pluginInstances;
    } catch (error) {
        throw new CLIError(error.message);
    }
};

const connectComponentConfig = async (
    component: ComponentConfig,
    componentConfigFile: ComponentConfigFile,
    plugins: ConnectPluginModule[]
): Promise<ConnectedComponent> => {
    const data: Data[] = [];
    if (plugins.length > 0) {
        // Execute all language plugins on the component if supports
        const pluginPromises = plugins.map(async plugin => {
            if (plugin.supports(component)) {
                const componentData = await plugin.process(component);

                data.push({
                    plugin: plugin.name,
                    ...removeEmptyFields(componentData)
                });
            }
        });

        await Promise.all(pluginPromises);
    }

    // TODO move urlPath preparation to service layer
    const urlPaths: Link[] = [];

    if (componentConfigFile.links) {
        componentConfigFile.links.forEach(link => {
            const { name, type, url } = link;
            const foundType = Object.keys(component).some(key => key === link.type);
            if (foundType) {
                const config = component[type];
                if (type === "storybook") {
                    const preparedUrls = prepareStorybookLinks(url, config);
                    preparedUrls.forEach(preparedUrl => {
                        urlPaths.push({ name, type, url: preparedUrl });
                    });
                } else if (type === "styleguidist") {
                    urlPaths.push({ name, type, url: urljoin(url, `#${encodeURIComponent(config.kind)}`) });
                } else {
                    urlPaths.push({ name, type: "custom", url: urljoin(url, config.urlPath) });
                }
            }
        });
    }

    if (componentConfigFile.github) {
        const branch = componentConfigFile.github.branch || defaults.github.branch;
        const url = componentConfigFile.github.url || defaults.github.url;
        const { repository } = componentConfigFile.github;
        urlPaths.push({
            type: "github",
            url: `${url}/${repository}/blob/${branch}/${component.path}`
        });
    }

    return {
        path: component.path,
        name: component.name,
        zeplinNames: component.zeplinNames,
        urlPaths,
        data
    };
};

const connectComponentConfigFile = async (
    componentConfigFile: ComponentConfigFile,
    connectPlugins: ConnectPluginModule[]
): Promise<ConnectedBarrelComponents> => {
    const connectedComponents = await Promise.all(
        componentConfigFile.components.map(component =>
            connectComponentConfig(component, componentConfigFile, connectPlugins)
        )
    );

    return {
        projects: componentConfigFile.projects || [],
        styleguides: componentConfigFile.styleguides || [],
        connectedComponents
    };
};

const connectComponentConfigFiles = (
    componentConfigFiles: ComponentConfigFile[],
    pluginModules: ConnectPluginModule[]
): Promise<ConnectedBarrelComponents[]> => {
    const promises = componentConfigFiles.map(componentConfigFile =>
        connectComponentConfigFile(componentConfigFile, pluginModules)
    );

    return Promise.all(promises);
};

export {
    importPlugins,
    connectComponentConfigFiles
};
