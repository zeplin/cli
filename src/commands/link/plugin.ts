import { ComponentConfig, ComponentData } from "link";
import { CLIError } from "../../errors";
import {
    LinkedComponent, Data, ComponentConfigFile, LinkedBarrelComponents, Url, ZeplinLinkPluginModule
} from "./interfaces";
import urljoin from "url-join";

interface ZeplinLinkPluginConstructor {
    new(): ZeplinLinkPluginModule;
}

// Helper method to initialize plugin classses
const constructLinkPlugin = (Constructor: ZeplinLinkPluginConstructor): ZeplinLinkPluginModule => new Constructor();

const removeEmptyFields = (componentData: ComponentData): ComponentData => {
    if (typeof componentData.description === "undefined" || componentData.description.trim() === "") {
        delete componentData.description;
    }

    if (typeof componentData.snippet === "undefined" || componentData.snippet.trim() === "") {
        delete componentData.snippet;
    }

    return componentData;
};

const importPlugins = async (plugins: string[]): Promise<ZeplinLinkPluginModule[]> => {
    try {
        const imports = plugins.map(async moduleName => {
            const linkPluginConstructor = (await import(moduleName)).default as ZeplinLinkPluginConstructor;

            const linkPluginInstance = constructLinkPlugin(linkPluginConstructor);
            linkPluginInstance.name = moduleName;
            return linkPluginInstance;
        });

        const pluginInstances = await Promise.all(imports);
        return pluginInstances;
    } catch (error) {
        throw new CLIError(error.message);
    }
};

const linkComponentConfig = async (
    component: ComponentConfig,
    baseURLs: Url[],
    plugins: ZeplinLinkPluginModule[]
): Promise<LinkedComponent> => {
    const data: Data[] = [];
    if (plugins.length > 0) {
        // Execute all language plugins on the component if supports
        const pluginPromises = plugins.map(async plugin => {
            if (plugin.supports(component)) {
                const componentData = await plugin.process(component);

                data.push({
                    plugin: plugin.name,
                    ...removeEmptyFields(componentData)
                });
            }
        });

        await Promise.all(pluginPromises);
    }

    const urlPaths: Url[] = [];
    if (component.urlPaths) {
        const configUrlPaths = component.urlPaths;
        Object.keys(configUrlPaths).forEach(type => {
            const baseUrl = baseURLs.find(u => u.type === type);
            const url = configUrlPaths[type];
            if (baseUrl) {
                urlPaths.push({ name: baseUrl.name, type, url: urljoin(baseUrl.url, url) });
            }
        });
    }

    return {
        path: component.path,
        name: component.name,
        zeplinNames: component.zeplinNames,
        urlPaths,
        data
    } as LinkedComponent;
};

const linkComponentConfigFile = async (
    componentConfigFile: ComponentConfigFile,
    linkPlugins: ZeplinLinkPluginModule[]
): Promise<LinkedBarrelComponents> => {
    const linkedComponents = await Promise.all(
        componentConfigFile.components.map(component =>
            linkComponentConfig(component, componentConfigFile.baseURLs, linkPlugins)
        )
    );

    return {
        projects: componentConfigFile.projects || [],
        styleguides: componentConfigFile.styleguides || [],
        connectedComponents: linkedComponents
    };
};

const linkComponentConfigFiles = (
    componentConfigFiles: ComponentConfigFile[],
    pluginModules: ZeplinLinkPluginModule[]
): Promise<LinkedBarrelComponents[]> => {
    const promises = componentConfigFiles.map(componentConfigFile =>
        linkComponentConfigFile(componentConfigFile, pluginModules)
    );

    return Promise.all(promises);
};

export {
    importPlugins,
    linkComponentConfigFiles
};
