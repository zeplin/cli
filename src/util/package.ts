import isYarnGlobal from "is-yarn-global";
import isInstalledGlobally from "is-installed-globally";
import hasYarn from "has-yarn";

const isRunningFromYarnGlobal = (): boolean => isYarnGlobal();

const isRunningFromGlobal = (): boolean => isRunningFromYarnGlobal() || isInstalledGlobally;

const projectHasYarn = (): boolean => hasYarn();

export {
    isRunningFromYarnGlobal,
    isRunningFromGlobal,
    projectHasYarn
};