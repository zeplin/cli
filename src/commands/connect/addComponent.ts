import chalk from "chalk";
import inquirer from "inquirer";
import dedent from "ts-dedent";
import {
    addComponentConfig,
    authentication,
    selectComponent,
    selectFile,
    selectResource,
    connectComponents
} from "../../tasks";
import { CliOptions, AddComponentContext } from "../../tasks/context/add-component";
import { Workflow } from "../../util/task";
import { indent } from "../../util/text";
import { AuthenticationService } from "../../service";
import { ConnectedComponentsService } from "./service";
import logger from "../../util/logger";
import { notInitialized, userSelectedNotToInitialize, addSummary } from "../../messages/initialize";
import { getComponentConfigFiles } from "./config";
import { initialize } from "..";
import { commandRunner } from "../../util/commander";

const askForInitialization = async (): Promise<boolean> => {
    const { confirmation } = await inquirer.prompt([{
        type: "confirm",
        message: "Do you want to initialize now?",
        default: false,
        name: "confirmation"
    }]);

    return confirmation;
};

export type AddComponentCommandOptions = CliOptions;

export async function addComponent(options: AddComponentCommandOptions): Promise<void> {
    try {
        const authService = new AuthenticationService();
        const connectService = new ConnectedComponentsService({ authService });

        const context: AddComponentContext = Object.assign(Object.create(null), {
            cliOptions: options
        }, {
            authService,
            connectService
        });

        const [existingConfigFile] = await getComponentConfigFiles([context.cliOptions.configFile]);
        if (!existingConfigFile) {
            logger.info(notInitialized());
            if (!(await askForInitialization())) {
                logger.info(userSelectedNotToInitialize());
                return;
            }

            return commandRunner(() => initialize(options))();
        }

        context.configFile = existingConfigFile;

        const workflow = new Workflow({
            context,
            tasks: [
                authentication,
                selectResource,
                selectComponent,
                selectFile,
                addComponentConfig,
                connectComponents
            ]
        });

        await workflow.run();

        logger.info(addSummary(context));
    } catch (error) {
        error.message = dedent`
            ${chalk.bold`Adding connected component failed.`}

            ${chalk.redBright(indent(error.message))}
        `;
        throw error;
    }
}
