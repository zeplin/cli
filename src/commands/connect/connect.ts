import chalk from "chalk";
import chokidar from "chokidar";
import dedent from "ts-dedent";
import logger from "../../util/logger";
import path from "path";
import { indent, stringify } from "../../util/text";
import { getComponentConfigFiles } from "./config";
import { ConnectedBarrelComponents } from "./interfaces/api";
import { connectComponentConfigFiles } from "./plugin";
import { ConnectDevServer } from "./server";
import { ConnectedComponentsService } from "./service";
import { isDefined } from "../../util/object";
import { flat } from "../../util/array";

const getComponentFilePaths = (connectedBarrels: ConnectedBarrelComponents[]): string[] => (
    flat(
        connectedBarrels.map(f =>
            f.items.map(c => (c.filePath ? path.resolve(c.filePath) : undefined)).filter(isDefined)
        )
    )
);

const generateConnectedComponents = async (
    options: Partial<Pick<ConnectOptions, "configFiles" | "plugins">>
): Promise<ConnectedBarrelComponents[]> => {
    const {
        configFiles,
        plugins
    } = options;

    const componentConfigFiles = await getComponentConfigFiles(configFiles, plugins);

    logger.debug(`component config files: ${stringify(componentConfigFiles)}`);

    const connectedBarrels = await connectComponentConfigFiles(componentConfigFiles);

    logger.debug(`connected barrels output: ${stringify(connectedBarrels)}`);

    return connectedBarrels;
};

const startDevServer = async (
    options: Pick<ConnectOptions, "configFiles" | "devModePort" | "devModeWatch" | "plugins">,
    connectedBarrels: ConnectedBarrelComponents[]
): Promise<void> => {
    const {
        configFiles,
        devModePort,
        devModeWatch,
        plugins
    } = options;

    logger.info("Starting development serverâ€¦");

    const devServer = new ConnectDevServer(connectedBarrels);

    if (devModeWatch) {
        let componentFiles = getComponentFilePaths(connectedBarrels);

        const watcher = chokidar.watch(
            [...configFiles, ...componentFiles],
            {
                cwd: process.cwd(),
                persistent: true,
                awaitWriteFinish: true
            }
        );

        watcher.on("change", async filePath => {
            logger.info((chalk.yellow(`\nFile change detected ${filePath}.\n`)));

            try {
                const updatedConnectedBarrels = await generateConnectedComponents({ configFiles, plugins });

                watcher.unwatch(componentFiles);

                devServer.updateConnectedBarrels(updatedConnectedBarrels);

                componentFiles = getComponentFilePaths(updatedConnectedBarrels);

                watcher.add(componentFiles);
            } catch (error) {
                logger.error(chalk.red(dedent`
                    Could not update Connected Components.
                    ${error}
                `));
            }
        });
    }

    await devServer.listen(devModePort);
};

const service = new ConnectedComponentsService();

const upload = async (connectedBarrels: ConnectedBarrelComponents[], { force }: ConnectOptions): Promise<void> => {
    logger.info("Connecting all Connected Components into Zeplinâ€¦");

    await service.uploadConnectedBarrels(connectedBarrels, { force });

    logger.info("ðŸ¦„ Components successfully connected to components in Zeplin.");
};

async function connect(options: ConnectOptions): Promise<void> {
    try {
        logger.debug(`connect options: ${stringify(options)}`);

        const connectedBarrels = await generateConnectedComponents(options);

        if (options.devMode) {
            await startDevServer(options, connectedBarrels);
        } else {
            await upload(connectedBarrels, options);
        }
    } catch (error) {
        error.message = dedent`
            ${chalk.bold`Connecting components to Zeplin components failed.`}

            ${chalk.redBright(indent(error.message))}
        `;
        throw error;
    }
}

export interface ConnectOptions {
    configFiles: string[];
    devMode: boolean;
    devModePort: number;
    devModeWatch: boolean;
    plugins: string[];
    force: boolean;
}

export {
    generateConnectedComponents,
    connect
};
