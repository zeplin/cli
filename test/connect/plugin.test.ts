import { mocked } from "ts-jest/utils";
import { connectComponentConfigFiles } from "../../src/commands/connect/plugin";
import * as samples from "../samples";
import * as packageUtils from "../../src/util/package";

jest.mock("../../src/util/package");

describe("PluginProcessor", () => {
    describe("when plugin not exists", () => {
        const { sampleConfig } = samples;
        sampleConfig.plugins = [{ name: "non-existent-plugin" }];
        afterEach(() => {
            jest.restoreAllMocks();
            jest.resetAllMocks();
        });

        describe("and CLI is globally installed in using yarn", () => {
            it("fails and prints instructions", async () => {
                mocked(packageUtils.isRunningFromYarnGlobal).mockReturnValue(true);
                mocked(packageUtils.isRunningFromGlobal).mockReturnValue(true);

                await expect(connectComponentConfigFiles([sampleConfig]))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe("and CLI is globally installed in using npm", () => {
            it("fails and prints instructions", async () => {
                mocked(packageUtils.isRunningFromGlobal).mockReturnValue(true);
                await expect(connectComponentConfigFiles([sampleConfig]))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe("and CLI is locally installed using yarn", () => {
            it("fails and prints instructions", async () => {
                mocked(packageUtils.projectHasYarn).mockReturnValue(true);

                await expect(connectComponentConfigFiles([sampleConfig]))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe("and CLI is locally installed using npm", () => {
            it("fails and prints instructions", async () => {
                await expect(connectComponentConfigFiles([sampleConfig]))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });
    });
});