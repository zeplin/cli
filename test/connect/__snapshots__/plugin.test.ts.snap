// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginProcessor when plugin not exists and CLI is globally installed in using npm fails and prints instructions 1`] = `
"Could not find plugin [1mnon-existent-plugin[22m failed.
Please make sure that it's globally installed and try again.
    npm install -g non-existent-plugin"
`;

exports[`PluginProcessor when plugin not exists and CLI is globally installed in using yarn fails and prints instructions 1`] = `
"Could not find plugin [1mnon-existent-plugin[22m failed.
Please make sure that it's globally installed and try again.
    yarn global add non-existent-plugin"
`;

exports[`PluginProcessor when plugin not exists and CLI is locally installed using npm fails and prints instructions 1`] = `
"Could not find plugin [1mnon-existent-plugin[22m failed.
Please make sure that it's installed and try again.
    npm install non-existent-plugin"
`;

exports[`PluginProcessor when plugin not exists and CLI is locally installed using yarn fails and prints instructions 1`] = `
"Could not find plugin [1mnon-existent-plugin[22m failed.
Please make sure that it's installed and try again.
    yarn add non-existent-plugin"
`;
